### ISSUE TITLE
[Improvement] Allow split huge data ShuffleDataFlushEvent to multiple small events

### ISSUE THREAD

**Issue by zuston** [apache/incubator-uniffle#398]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-uniffle/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

[Improvement] Allow split huge data ShuffleDataFlushEvent to multiple small events.

As the problem mentioned by https://github.com/apache/incubator-uniffle/issues/378#issuecomment-1344260350, the small event will benefit more after the PR of #396 is applied.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

**Comment by zuston** [comment]
cc @jerqi

**Comment by jerqi** [comment]
Could we avoid huge event by using this pr https://github.com/apache/incubator-uniffle/pull/176 ?

**Comment by zuston** [comment]
> Could we avoid huge event by using this pr #176 ?

Yes. But I don't hope so, detailed reason could be found in https://github.com/apache/incubator-uniffle/issues/378#issuecomment-1342637715 of second case.

**Comment by jerqi** [comment]
> > Could we avoid huge event by using this pr #176 ?
> 
> Yes. But I don't hope so, detailed reason could be found in [#378 (comment)](https://github.com/apache/incubator-uniffle/issues/378#issuecomment-1342637715) of second case.

I can't get your point.

**Comment by zuston** [comment]
If the single buffer flush mechanism is enabled, it will be applied all partition data. But when the memory is free and the concurrent apps number is small, there is unnecessary to flush data to localfile or disk. In my view, it will make the local disk free if the single buffer flush size  > cold storage threshold size. And it will make memory free  if the single buffer flush size  < cold storage threshold size. 

Overall, this mechanism will not make full use of memory and then leads to performance regression.

**Comment by zuston** [comment]
> Because if we write data to storage, and then we read the data quickly, the data may be in the page cache.

But in the most cases, the read wont happen quickly after writing, especially for many tasks.

**Comment by jerqi** [comment]
> > Because if we write data to storage, and then we read the data quickly, the data may be in the page cache.
> 
> But in the most cases, the read wont happen quickly after writing, especially for many tasks.

Maybe we should some tests prove the effect.

**Comment by zuston** [comment]
I have tested in our online cluster https://github.com/apache/incubator-uniffle/issues/378#issuecomment-1342637715.

**Comment by zuston** [comment]
It's still useless after #396. Single huge event only can be handled by single thread to flush to HDFS, which don't benefit from concurrent written mechanism.

**Comment by zuston** [comment]
> If you use single buffer limit at the same time, you will benefit the feature.

I know this, but I don't hope all large buffer flush to persist storage, especially for having enough free memory.
